File: ImageToPDFConverter -.py
import os
from tkinter import Tk, filedialog, messagebox, Button
from PIL import Image
from fpdf import FPDF


class PdfMaker:
    def __init__(self, root):
        self.root = root
        self.root.title("PDF Maker")
        self.images = []  # List to store selected images

        # Buttons
        Button(root, text="Select Images", command=self.select_images).pack(pady=10)
        Button(root, text="Create PDF", command=self.create_pdf).pack(pady=10)

    def select_images(self):
        files = filedialog.askopenfilenames(filetypes=[("Image Files", "*.jpg *.jpeg *.png *.bmp *.gif")])
        if not files:
            return

        for file in files:
            try:
                img = Image.open(file)
                self.images.append(img)
            except Exception as e:
                messagebox.showerror("Error", f"An error occurred while loading the image: {e}")

        messagebox.showinfo("Success", "Images have been added!")

    def create_pdf(self):
        if not self.images:
            messagebox.showerror("Error", "Please select at least one image.")
            return

        pdf_path = filedialog.asksaveasfilename(defaultextension=".pdf", filetypes=[("PDF Files", "*.pdf")])
        if not pdf_path:
            return

        pdf = FPDF(unit="pt", format="A4")
        try:
            for img in self.images:
                pdf.add_page()
                img_width, img_height = img.size
                pdf_width, pdf_height = 595.28, 841.89  # A4 size in points (pt)

                scale_factor = min(pdf_width / img_width, pdf_height / img_height)
                scaled_width = img_width * scale_factor
                scaled_height = img_height * scale_factor
                x = (pdf_width - scaled_width) / 2
                y = (pdf_height - scaled_height) / 2

                img_path = "temp_image.jpg"
                img.save(img_path)
                pdf.image(img_path, x=x, y=y, w=scaled_width, h=scaled_height)
                os.remove(img_path)

            pdf.output(pdf_path)
            messagebox.showinfo("Success", "PDF successfully created!")
        except Exception as e:
            messagebox.showerror("Error", f"An error occurred while creating the PDF: {e}")


if __name__ == "__main__":
    root = Tk()
    root.geometry("300x150")
    app = PdfMaker(root)
    root.mainloop()

File: ImageToPDFConverter - .cpp
#include <QApplication>
#include <QWidget>
#include <QPushButton>
#include <QFileDialog>
#include <QMessageBox>
#include <QImage>
#include <QPainter>
#include <QPdfWriter>

class PdfMaker : public QWidget
{
    Q_OBJECT

public:
    PdfMaker(QWidget *parent = nullptr) : QWidget(parent)
    {
        // Button to select images
        QPushButton *selectButton = new QPushButton("Select Images", this);
        selectButton->setGeometry(50, 50, 200, 40);
        connect(selectButton, &QPushButton::clicked, this, &PdfMaker::selectImages);

        // Button to create PDF
        QPushButton *createButton = new QPushButton("Create PDF", this);
        createButton->setGeometry(50, 100, 200, 40);
        connect(createButton, &QPushButton::clicked, this, &PdfMaker::createPdf);
    }

private:
    QStringList imagePaths; // List to store paths of selected images

    void selectImages()
    {
        imagePaths = QFileDialog::getOpenFileNames(this, "Select Images", QString(), "Image Files (*.jpg *.jpeg *.png *.bmp *.gif)");
        if (imagePaths.isEmpty())
        {
            QMessageBox::warning(this, "Error", "No images selected.");
        }
        else
        {
            QMessageBox::information(this, "Success", "Images have been added!");
        }
    }

    void createPdf()
    {
        if (imagePaths.isEmpty())
        {
            QMessageBox::warning(this, "Error", "Please select at least one image.");
            return;
        }

        QString pdfPath = QFileDialog::getSaveFileName(this, "Create PDF", QString(), "PDF Files (*.pdf)");
        if (pdfPath.isEmpty())
            return;

        QPdfWriter writer(pdfPath);
        writer.setPageSize(QPagedPaintDevice::A4);
        QPainter painter(&writer);

        try
        {
            for (const QString &imagePath : imagePaths)
            {
                QImage image(imagePath);
                if (image.isNull())
                {
                    throw std::runtime_error("Failed to load image.");
                }

                QRect pageRect = writer.pageRect();
                double scaleFactor = qMin(double(pageRect.width()) / image.width(), double(pageRect.height()) / image.height());
                int scaledWidth = image.width() * scaleFactor;
                int scaledHeight = image.height() * scaleFactor;
                int x = (pageRect.width() - scaledWidth) / 2;
                int y = (pageRect.height() - scaledHeight) / 2;

                painter.drawImage(QRect(x, y, scaledWidth, scaledHeight), image);
                writer.newPage();
            }

            QMessageBox::information(this, "Success", "PDF successfully created!");
        }
        catch (const std::exception &e)
        {
            QMessageBox::critical(this, "Error", e.what());
        }

        painter.end();
    }
};

int main(int argc, char *argv[])
{
    QApplication app(argc, argv);

    PdfMaker window;
    window.resize(300, 200);
    window.setWindowTitle("PDF Maker");
    window.show();

    return app.exec();
}
